cmd/kubeadm/app/util/config/masterconfig.go

```diff
@@ -126,15 +138,54 @@ func ConfigFileAndDefaultsToInternalConfig(cfgPath string, defaultversionedcfg *
 // BytesToInternalConfig converts a byte array to an internal, defaulted and validated configuration object
 func BytesToInternalConfig(b []byte) (*kubeadmapi.MasterConfiguration, error) {
 	internalcfg := &kubeadmapi.MasterConfiguration{}
+	kubeletConfig := &kubeletconfigv1beta1.KubeletConfiguration{}
+	kubeProxyConfig := &kubeproxyconfigv1alpha1.KubeProxyConfiguration{}
+	kubeletConfigLoaded, kubeProxyConfigLoaded := false, false
 
 	if err := DetectUnsupportedVersion(b); err != nil {
 		return nil, err
 	}
 
-	if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), b, internalcfg); err != nil {
+	gvkmap, err := kubeadmutil.SplitYAMLDocuments(b)
+	if err != nil {
 		return nil, err
 	}
 
+	for gvk, fileContent := range gvkmap {
+
+		switch gvk.Kind {
+		case "KubeletConfiguration":
+			_, kubeletCodecs, err := kubeletscheme.NewSchemeAndCodecs()
+			if err != nil {
+				return nil, err
+			}
+			if err := runtime.DecodeInto(kubeletCodecs.UniversalDecoder(), fileContent, kubeletConfig); err != nil {
+				return nil, err
+			}
+			kubeletConfig.TypeMeta = metav1.TypeMeta{}
+			kubeletConfigLoaded = true
+		case "KubeProxyConfiguration":
+			if err := runtime.DecodeInto(kubeproxyscheme.Codecs.UniversalDecoder(), fileContent, kubeProxyConfig); err != nil {
+				return nil, err
+			}
+			kubeProxyConfig.TypeMeta = metav1.TypeMeta{}
+			kubeProxyConfigLoaded = true
+		case kubeadmconstants.MasterConfigurationKind:
+			if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), fileContent, internalcfg); err != nil {
+				return nil, err
+			}
+		default:
+			fmt.Printf("[config] WARNING: Ignored YAML document with GroupVersionKind %v", gvk)
+		}
+	}
+	// Set the componentconfigs on the internalcfg object; ONLY if they have been loaded from an external YAML file
+	if kubeletConfigLoaded {
+		internalcfg.ComponentConfigs.Kubelet = kubeletConfig
```

Comment:
What about moving this line above at the end of `case "KubeletConfiguration"` (~line 166), so we can avoid the `kubeletConfigLoaded` flag?

Reply:
This code has been rewritten